File: app/src/main/java/com/nfcalarmclock/NacMissedAlarmNotification.java
package com.nfcalarmclock;

import android.annotation.TargetApi;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.os.Build;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;

/**
 */
public class NacMissedAlarmNotification
	extends NacNotification
{

	/**
	 * Channel.
	 */
	protected static final String MISSED_CHANNEL = "NacNotiChannelMissed";

	/**
	 * Group.
	 */
	protected static final String MISSED_GROUP = "NacNotiGroupMissed";

	/**
	 * Notification ID.
	 */
	protected static final int MISSED_ID = 222;

	/**
	 * List of lines to show in the notification.
	 */
	protected List<CharSequence> mLines;

	/**
	 */
	public NacMissedAlarmNotification()
	{
		super();
	}

	/**
	 */
	public NacMissedAlarmNotification(Context context)
	{
		super(context);
		this.mLines = null;
	}

	/**
	 */
	@Override
	public void createChannel()
	{
		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O)
		{
			return;
		}

		Context context = this.getContext();
		NacSharedConstants cons = new NacSharedConstants(context);
		NotificationChannel channel = new NotificationChannel(MISSED_CHANNEL,
			cons.getMissedNotification(),
			NotificationManager.IMPORTANCE_DEFAULT);
		NotificationManager manager = context.getSystemService(
			NotificationManager.class);

		channel.setDescription(cons.getDescriptionMissedNotification());
		channel.setShowBadge(true);
		channel.enableLights(true);
		channel.enableVibration(true);
		channel.setLockscreenVisibility(Notification.VISIBILITY_PRIVATE);
		manager.createNotificationChannel(channel);
	}

	/**
	 * @return The notification group.
	 */
	public String getGroup()
	{
		return MISSED_GROUP;
	}

	/**
	 * @return The list of lines in the notification.
	 */
	protected List<CharSequence> getLines()
	{
		return this.mLines;
	}

	/**
	 * @return The notification text.
	 */
	public String getText(NacAlarm alarm)
	{
		Context context = this.getContext();
		String time = alarm.getFullTime(context);
		String name = alarm.getName();
		Locale locale = Locale.getDefault();

		return (name.isEmpty()) ? time
			: String.format(locale, "%1$s  —  %2$s", time, name);
	}

	/**
	 * @return The notification title.
	 */
	public String getTitle()
	{
		Context context = this.getContext();
		List<CharSequence> lines = this.getLines();
		NacSharedConstants cons = new NacSharedConstants(context);
		int count = (lines == null) ? 1 : lines.size();

		return cons.getMissedAlarm(count);
	}

	/**
	 * @see hide
	 */
	public void hide()
	{
		Context context = this.getContext();
		NotificationManagerCompat manager = this.getNotificationManager(context);
		manager.cancel(MISSED_ID);
	}

	/**
	 * Show the notification.
	 */
	@TargetApi(Build.VERSION_CODES.M)
	public void show(NacAlarm alarm)
	{
		Context context = this.getContext();
		NotificationManager manager = (NotificationManager)
			context.getSystemService(Context.NOTIFICATION_SERVICE);

		if (alarm == null)
		{
			//manager.cancel(MISSED_ID);
			return;
		}


		this.mLines = this.getNotificationText(manager, alarm);
		List<CharSequence> lines = this.mLines;
		NotificationCompat.InboxStyle inbox = new NotificationCompat
			.InboxStyle();
		NotificationCompat.Builder builder = this.build(MISSED_CHANNEL)
			.setAutoCancel(true)
			.setShowWhen(true);
		int size = lines.size();

		for (CharSequence l : lines)
		{
			inbox.addLine(l);
		}

		if (size == 0)
		{
			manager.cancel(MISSED_ID);
		}
		else if (size > 0)
		{
			NacSharedConstants cons = new NacSharedConstants(context);
			Locale locale = Locale.getDefault();
			String word = cons.getAlarm(size);
			String content = String.format(locale, "%1$d %2$s", size, word);

			builder.setContentText(content);
		}

		builder.setStyle(inbox);
		manager.notify(MISSED_ID, builder.build());
	}

}

File: app/src/main/kotlin/com/nfcalarmclock/alarm/options/missedalarm/NacMissedAlarmNotification.kt
package com.nfcalarmclock.alarm.options.missedalarm

import android.app.NotificationChannel
import android.app.PendingIntent
import android.content.Context
import android.os.Build
import androidx.annotation.RequiresApi
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import com.nfcalarmclock.R
import com.nfcalarmclock.alarm.db.NacAlarm
import com.nfcalarmclock.main.NacMainActivity
import com.nfcalarmclock.system.NacCalendar
import com.nfcalarmclock.view.notification.NacBaseNotificationBuilder
import java.util.Calendar

/**
 * Missed alarm notification.
 *
 * @param context Context.
 * @param alarm Alarm.
 */
class NacMissedAlarmNotification(
	context: Context,
	private val alarm: NacAlarm
) : NacBaseNotificationBuilder(context, "NacNotiChannelMissed")
{

	/**
	 * @see NacBaseNotificationBuilder.id
	 */
	override val id: Int
		get() = BASE_ID + alarm.id.toInt()

	/**
	 * @see NacBaseNotificationBuilder.channelName
	 */
	override val channelName: String = context.getString(R.string.title_missed_alarms)

	/**
	 * @see NacBaseNotificationBuilder.channelDescription
	 */
	override val channelDescription: String = context.getString(R.string.description_missed_alarm)

	/**
	 * @see NacBaseNotificationBuilder.priorityLevel
	 */
	override val priorityLevel: Int = NotificationCompat.PRIORITY_DEFAULT

	/**
	 * @see NacBaseNotificationBuilder.channelImportance
	 */
	override val channelImportance: Int = NotificationManagerCompat.IMPORTANCE_DEFAULT

	/**
	 * @see NacBaseNotificationBuilder.group
	 */
	override val group: String = "NacNotiGroupMissed"

	/**
	 * @see NacBaseNotificationBuilder.contentText
	 */
	override val contentText: String
		get()
		{
			// Calendar of the alarm. If it is in the future, subtract a day
			val now = Calendar.getInstance()
			val cal = NacCalendar.alarmToCalendar(alarm)

			if (cal > now)
			{
				cal.add(Calendar.DAY_OF_MONTH, -1)
			}

			// Time of the alarm
			val time = NacCalendar.getFullTime(context, cal)

			// Notification text
			return if (alarm.name.isEmpty())
			{
				time
			}
			else
			{
				"$time  —  ${alarm.name}"
			}
		}

	/**
	 * @see NacBaseNotificationBuilder.contentPendingIntent
	 */
	override val contentPendingIntent: PendingIntent
		get() = NacMainActivity.getStartPendingIntent(context)

	/**
	 * Constructor.
	 */
	init
	{
		// Create the channel
		setupChannel()

		// Get the title
		val title = context.resources.getString(R.string.title_missed_alarm)

		// Build the notification
		this.setPriority(priorityLevel)
			.setCategory(category)
			.setGroup(group)
			.setContentTitle(title)
			.setContentText(contentText)
			.setContentIntent(contentPendingIntent)
			.setSmallIcon(smallIcon)
			.setTicker(channelName)
			.setColor(ContextCompat.getColor(context, R.color.ic_launcher_background))
			.setGroupSummary(true)
			.setAutoCancel(true)
			.setShowWhen(true)
	}

	/**
	 * @see NacBaseNotificationBuilder.createChannel
	 */
	@RequiresApi(Build.VERSION_CODES.O)
	override fun createChannel(): NotificationChannel
	{
		// Create the channel
		val channel = super.createChannel()

		// Setup the channel
		channel.setShowBadge(true)
		channel.enableLights(true)
		channel.enableVibration(true)

		return channel
	}

	companion object
	{

		/**
		 * The base ID value to use for this type of notification.
		 */
		const val BASE_ID: Int = 222

	}

}